homeassistant/components/sensor/elkm1.py

```diff
@@ -0,0 +1,315 @@
+"""
+Support for control of ElkM1 sensors.
+
+On the ElkM1 there are 6 types of sensors:
+- Counters that are integers that can be read
+- Keypads that have temperature (not all models, but no way to know)
+- Panel that monitors the state of the connection, versions, etc.
+- Settings that are used to trigger Elk-M1 automations
+- Thermostats that have temperature
+- Zones that are multi-state, voltage, or temperature.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.elkm1/
+"""
+
+import voluptuous as vol
+
+from homeassistant.components.elkm1 import (
+    DOMAIN as ELK_DOMAIN, create_elk_entities, ElkEntity)
+from homeassistant.components.sensor import DOMAIN
+from homeassistant.const import ATTR_ENTITY_ID
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect, async_dispatcher_send)
+
+DEPENDENCIES = [ELK_DOMAIN]
+
+SPEAK_SERVICE_SCHEMA = vol.Schema({
+    vol.Optional(ATTR_ENTITY_ID): cv.entity_ids,
+    vol.Required('number'): vol.Range(min=0, max=999),
+})
+
+SIGNAL_SPEAK_WORD = 'elkm1_speak_word'
+SIGNAL_SPEAK_PHRASE = 'elkm1_speak_phrase'
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Create the Elk-M1 sensor platform."""
+    if discovery_info is None:
+        return
+
+    elk = hass.data[ELK_DOMAIN]['elk']
+    entities = create_elk_entities(
+        hass, elk.counters, 'counter', ElkCounter, [])
+    entities = create_elk_entities(
+        hass, elk.keypads, 'keypad', ElkKeypad, entities)
+    entities = create_elk_entities(
+        hass, [elk.panel], 'panel', ElkPanel, entities)
+    entities = create_elk_entities(
+        hass, elk.settings, 'setting', ElkSetting, entities)
+    entities = create_elk_entities(
+        hass, elk.thermostats, 'thermostat', ElkThermostat, entities)
+    entities = create_elk_entities(
+        hass, elk.zones, 'zone', ElkZone, entities)
+    async_add_entities(entities, True)
+
+    def _dispatch(signal, entity_ids, *args):
+        for entity_id in entity_ids:
+            async_dispatcher_send(
+                hass, '{}_{}'.format(signal, entity_id), *args)
+
+    def _speak_word_service(service):
+        entity_ids = service.data.get(ATTR_ENTITY_ID, [])
+        args = (service.data.get('number'))
+        _dispatch(SIGNAL_SPEAK_WORD, entity_ids, *args)
+
+    def _speak_phrase_service(service):
+        entity_ids = service.data.get(ATTR_ENTITY_ID, [])
+        args = (service.data.get('number'))
+        _dispatch(SIGNAL_SPEAK_PHRASE, entity_ids, *args)
+
+    hass.services.async_register(DOMAIN, 'elkm1_sensor_speak_word',
```

Comment:
Sensors are up until now read only, so we have never implemented any services for these platforms. Probably move the services to the elkm1 component. This is an architecture question.

I don't really see why these services belong to the sensor platform in the first place. "speak word" and "speak phrase" doesn't sound like something a sensor would do. More like a speaker. But I don't know this integration.

Reply:
I can move the services. They seemed to be a better fit against the alarm panel rather than the entire config but serious don't know or care much as long as they can be easily invoked.

Reply:
I suggest moving them to the component.

Reply:
When you say move to the component, is that `elkm1/__init__.py` `async_setup`? Would I register the service as `elkm1.speak_word`, for example?

Reply:
Yes!